api/src/main/java/jakarta/servlet/ServletContext.java

```diff
@@ -1436,4 +1523,97 @@
+     * character.</li>
+     * <li>a 'path parameter' is the part of a URI segment as defined by the obsoleted
+     * <a href="https://datatracker.ietf.org/doc/html/rfc2396#section-3.3">RFC2396 Section 3.3</a>.</li>
+     * </ul>
+     * 
+     * @since Servlet 6.0
+     */
+    public enum PathCodingMode {
+        /**
+         * Effected methods will return/accept paths as was done prior to the introduction of the PathCodingMode in Servlet 6.0.
+         * The path returned by {@link HttpServletRequest#getContextPath()} is not decoded. Paths returned by the methods
+         * {@link HttpServletRequest#getServletPath()} and {@link HttpServletRequest#getPathInfo()} are fully decoded and have
+         * all path parameters removed. Decoding may include ambiguous characters such as a decoded <code>%2F</code>, which is
+         * indistinguishable from a segment separator '/' in the request URI. Methods accepting paths may accept encoded
+         * characters. A '?' character in a path will be treated as indicating a query string and will not be encoded. Some
+         * implementation specific behavioural differences may be apparent in this mode.
+         */
+        LEGACY,
+
+        /**
+         * Effected methods will only return/accept paths that do not contain any encoded reserved characters and with all path
+         * parameters removed. If the requested URI contains encoded reserved characters, then a call to a method that returns
+         * such a path will throw {@link IllegalStateException}.
+         *
+         * <p>
+         * TODO pick between this: If a path with encoded reserve characters is passed to a method taking a path as a parameter,
+         * then it will throw {@link MalformedURLException} if declared. otherwise a {@link IllegalArgumentException}.
+         *
+         * <p>
+         * TODO or this: If a path with encoded reserve characters is received, then the request will be rejected by the servlet
+         * container with a 400 BAD REQUEST response.
+         */
+        DECODED,
+
+        /**
+         * Effected methods will return/accept paths in which only reserved characters are encoded. Methods returning paths will
+         * decode all non-reserved characters and remove only the <code>jsessionid</code> path parameter. All other path
+         * parameters will be returned and considered part of the path for the purposes of path matching. Methods that take
+         * paths expect the reserved characters to be encoded if they are to be considered part of a path segment, otherwise
+         * they will not be encoded and will have their reserved interpretation if possible. If a valid interpretation of a
+         * non-encoded reserved character is not possible, then the method will throw {@link MalformedURLException} if declared,
+         * otherwise a {@link IllegalArgumentException}.
+         */
+        ENCODED,
```

Comment:
If path parameters may be present in any segment present then it is not going to be possible write a `<url-pattern>` that matches the servlet path. The only place path parameters will be usable is in path info. I think that is too restrictive.

I currently favour a mode that maps requests to servlets based using the simplified version of `DECODED` that I described above but for methods that return path, those paths include the path parameters as described here. The mapped path (i.e. the one without the path parameters) remains available via `HttpServletMapping`.

The idea is that with this approach things are essentially unchanged for the majority of applications. The small number of applications using path parameters may need changes. Containers have the option of providing a `LEGACY` mode if there is sufficient demand from their user base from users with applications that use path parameters. I suspect that given that most apps are going to migrate from Jakarta EE 8 to Jakarta EE 10 then the few apps where changes to path parameters have an impact will recode the relevant bits of their app to handle the changes.

Reply:
I don't like treating ';' as a normal character for path methods, but special for path matching.    I think it needs to be one or the other, not both... specially when we consider the complexities of methods taking paths.

We could continue to treat ';' as special and remove all parameters from path methods. Apps wanting to handle them would need to use the requestURI as they do now.   But we need to specify handling of scary URIs that use ';' to hide '..' etc.

Or we treat ';' as a character allowed in a segment (as is now apparently is). Then `/..;` is not a problem because the segment is `..;` and not `..`.   But if we are treating is as part of a segment, the it needs to be able to be mapped.   I don't see it is impossible to map that because ';' can be included in a `<url-pattern>` as the spec says that all other parts of a pattern string are used for exact matches, so they can contain `;`, `\`, `?` and other scary characters.

Reply:
Sorry - forgot to click the add button for this comment.

Arrgghhhh! Just when I think there is a workable solution in sight, I realise I have forgotten about one of the requirements. I think the suggestions above would give us a workable solution that removes most (hopefully all) the ambiguity in the current API. What it doesn't address is providing access to the path parameters and/or the 'raw' versions of contextPath, servletPath and pathInfo.

What if, rather than introducing the concept of modes, we added methods to `HttpServletMapping`. I'm thinking:
- `getContextPathEncoded()`
- `getServletPathEncoded()`
- `getPathInfoEncoded()`

where only the following had been performed;
- extract query string, if any, for separate processing
- decode unreserved %nn
- normalize
- reject dot-segemts with path parameters